File: semestre-1/algoritmos/MR-Alg-05/MR-Alg-5-Ex-05.py
from math import log10, floor

# A função ordinal() se baseia nos valores adicionados nesta lista, onde cada posição na lista equivale
# aos números ordinais da casa de unidade, dezena, centena, etc. Atualmente, só aceita até centena (1-999),
# mas pode ser facilmente expandida se inserir os próximos nomes (1000, 2000, 3000, ..., 9000)
ordinais = [
        [       # De 1 a 9
            "", # Ignora posição zero
            "primeiro",
            "segundo",
            "terceiro",
            "quarto",
            "quinto",
            "sexto",
            "sétimo",
            "oitavo",
            "nono",
        ],
        [       # De 10 a 90
            "", # Ignora posição zero
            "décimo",
            "vigésimo",
            "trigésimo",
            "quadrigésimo",
            "quinquagésimo",
            "sexagésimo",
            "septagésimo",
            "octagésimo",
            "nonagésimo",
        ],  
        [       # De 100 a 900
            "", # Ignora posição zero
            "centésimo",
            "ducentésimo",
            "tricentésimo",
            "quadringentésimo",
            "quingentésimo",
            "sexcentésimo",
            "septingentésimo",
            "octingentésimo",
            "noningentésimo"
        ]
    ]

def ordinal(n):
    # Quantidade de digitos no número
    digitos = floor(log10(n)) + 1

    # Verifica se está implementada a conversão de cardinal para decimal para o tanto
    # de dígitos do número dado (ou se é um número negativo)
    if len(ordinais) < digitos or n <= 0:
        return ""
    
    # Inicialize numero ordinal
    ord_nomes = []

    for i in range(1, digitos + 1):
        # Pega o dígito dessa casa
        digito = n // 10 ** (digitos - i) % 10

        # Se for zero, ignora
        if digito == 0: continue

        # Adiciona o nome em ordinal do dígito atual
        ord_nomes.append(ordinais[digitos - i][digito])

    # Separa os nomes com espaço
    return " ".join(ord_nomes)

def main():
    for i in range(1, 13):
        print(f"{i}° = '{ordinal(i)}'")

    final = 10**len(ordinais)
    if input(f"\nDeseja mostrar os ordinais de 13 a {final-1}? (S/N): ").lower() == "s":
        for i in range(13, final):
            print(f"{i}° = '{ordinal(i)}'")

if __name__ == "__main__":
    main()

File: semestre-1/algoritmos/MR-Alg-05/MR-Alg-5-Ex.05.c
#include "stdio.h"
#include "stdlib.h"
#include "math.h"
#include <string.h>

char* ordinal(int n) {
    const char* ordinais[][10] = {
        {
            "",
            "primeiro",
            "segundo",
            "terceiro",
            "quarto",
            "quinto",
            "sexto",
            "setimo",
            "oitavo",
            "nono",
        },
        {
            "",
            "decimo",
            "vigesimo",
            "trigesimo",
            "quadrigesimo",
            "quinquagesimo",
            "sexagesimo",
            "septagesimo",
            "octagesimo",
            "nonagesimo",
        },
        {
            "",
            "centesimo",
            "ducentesimo",
            "tricentesimo",
            "quadringentesimo",
            "quingentesimo",
            "sexcentesimo",
            "septingentesimo",
            "octingentesimo",
            "noningentesimo"
        }
    };
    size_t size = sizeof(ordinais) / sizeof(ordinais[0]);
    int digitos = floor(log10((double) n)) + 1;

    if (digitos > size || n < 0) {
        return "Not implemented";
    }
    
    char* str = malloc(sizeof(char) * 41);
    str[0] = '\0';

    for (int i = 1; i <= digitos; i++) {
        int digito = (int)floor(n / pow(10, digitos -i)) % 10;

        if (digito == 0) continue;

        if (i != 1) {
            strcat(str, " ");
            strcat(str, ordinais[digitos - i][digito]);
        } else {
            strcat(str, ordinais[digitos - i][digito]);
        }
    }

    return str;
}

void main() {
    for (int i = 1; i <= 12; i++) {
        const char* ord = ordinal(i);
        printf("%d = %s\n", i, ord);
    }

    char ans[2];
    printf("Show ordinals from 13 to 999 ? (Y/N): ");
    scanf("%c", &ans);

    if (strcmp(ans, "y") == 0 || strcmp(ans, "Y") == 0) {
        for (int i = 13; i <= 999; i++) {
            const char* ord = ordinal(i);
            printf("%d = %s\n", i, ord);
        }
    }
}